ipfs-9-to-10/migration/migration.go

```diff
@@ -0,0 +1,120 @@
+package mg9
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	migrate "github.com/ipfs/fs-repo-migrations/go-migrate"
+	lock "github.com/ipfs/fs-repo-migrations/ipfs-1-to-2/repolock"
+	mfsr "github.com/ipfs/fs-repo-migrations/mfsr"
+	log "github.com/ipfs/fs-repo-migrations/stump"
+)
+
+type Migration struct{}
+
+func (m Migration) Versions() string {
+	return "9-to-10"
+}
+
+func (m Migration) Reversible() bool {
+	return true
+}
+
+func (m Migration) Apply(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("applying %s repo migration", m.Versions())
+
+	log.VLog("locking repo at %q", opts.Path)
+	lk, err := lock.Lock2(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+
+	log.VLog("  - verifying version is '9'")
+	if err := repo.CheckVersion("9"); err != nil {
+		return err
+	}
+
+	log.Log("> Upgrading config to new format")
+
+	path := filepath.Join(opts.Path, "config")
+	if err := convertFile(path, true, ver9to10Bootstrap, ver9to10Swarm); err != nil {
+		return err
+	}
+
+	if err := repo.WriteVersion("10"); err != nil {
+		log.Error("failed to update version file to 10")
+		return err
+	}
+
+	log.Log("updated version file")
+
+	return nil
+}
+
+func writePhase(file string, phase int) error {
+	return ioutil.WriteFile(file, []byte(fmt.Sprint(phase)), 0666)
+}
+
+func readPhase(file string) (int, error) {
+	data, err := ioutil.ReadFile(file)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return 0, nil
+		}
+		return 0, err
+	}
+
+	return strconv.Atoi(string(data))
+}
+
+func (m Migration) Revert(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("reverting migration")
+	lk, err := lock.Lock2(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+	if err := repo.CheckVersion("10"); err != nil {
+		return err
+	}
+
+	phasefile := filepath.Join(opts.Path, "revert-phase")
+	path := filepath.Join(opts.Path, "config")
+
+	phase, err := readPhase(phasefile)
+	if err != nil {
+		return fmt.Errorf("reading revert phase: %s", err)
+	}
+
+	defer os.Remove(phasefile)
+	for ; phase < 2; phase++ {
+		switch phase {
+		case 0:
+			if err := convertFile(path, false, ver10to9Bootstrap, ver10to9Swarm); err != nil {
```

Comment:
We don't really need to do anything in the reverse migration. The QUIC addresses will be ignored if we don't support them.

Can we even correctly implement the reverse migration?